rosplan_knowledge_base/src/PDDLKnowledgeBase.cpp

```diff
@@ -3,204 +3,224 @@
                 if(sensed_predicates.find(req.name) == sensed_predicates.end()) {
                     sensed_predicates[req.name] = false;
                 }
                 res.is_sensed = sensed_predicates[req.name];
 
-				return true;
-			}
-		}
-		return false;
-	}
-
-	/*-------------------------------------*/
-	/* add initial state to knowledge base */
-	/*-------------------------------------*/
-
-	/* get the initial state from the domain and problem files */
-	void PDDLKnowledgeBase::addInitialState() {
-		VALVisitorProblem problem_visitor(domain_parser.domain, problem_parser.problem);
-		model_instances = problem_visitor.returnInstances();
-		model_facts = problem_visitor.returnFacts();
-		model_functions = problem_visitor.returnFunctions();
-		model_goals = problem_visitor.returnGoals();
-		model_timed_initial_literals = problem_visitor.returnTimedKnowledge();
-		if (problem_parser.problem->metric) {
-			model_metric = problem_visitor.returnMetric();
-		}
-	}
-
-	void PDDLKnowledgeBase::parseDomain(const std::string &domain_file_path, const std::string &problem_file_path) {
-		std::ifstream file_check;
-		VAL1_2::analysis* an_analysis = new VAL1_2::analysis;
-		domain_parser.val_analysis = an_analysis;
-		problem_parser.val_analysis = an_analysis;
-
-		// parse domain
-		ROS_INFO("KCL: (%s) Parsing domain", ros::this_node::getName().c_str());
-		domain_parser.domain_parsed = false;
-		VAL1_2::domain* domain;
-		file_check.open(domain_file_path.c_str());
-		if(!file_check.good()) {
-			ROS_ERROR("KCL: (%s) Domain file does not exist.", ros::this_node::getName().c_str());
-			ros::shutdown();
-		} else {
-			file_check.close();
-			domain = domain_parser.parseDomain(domain_file_path);
-			if(!domain) {
-				ROS_ERROR("KCL: (%s) There were syntax errors in the domain file.", ros::this_node::getName().c_str());
-				ros::shutdown();
-			}
-		}
-
-		// parse problem and add initial state
-		if(problem_file_path != "") {
-			ROS_INFO("KCL: (%s) Parsing initial state", ros::this_node::getName().c_str());
-			problem_parser.problem_parsed = false;
-			file_check.open(problem_file_path.c_str());
-			if(!file_check.good()) {
-				ROS_WARN("KCL: (%s) Initial state file does not exist.", ros::this_node::getName().c_str());
-			} else {
-				file_check.close();
-				VAL1_2::problem* problem = problem_parser.parseProblem(problem_file_path);
-				if(problem) {
-					addInitialState();
-				} else {
-					ROS_WARN("KCL: (%s) There were syntax errors in the problem file.", ros::this_node::getName().c_str());
-				}
-			}
-		}
-	}
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /*-------------------------------------*/
+    /* add initial state to knowledge base */
+    /*-------------------------------------*/
+
+    /* get constants from the domain file */
+    void PDDLKnowledgeBase::addConstants() {
+
+        VAL1_2::domain* domain;
+        domain = domain_parser.parseDomain("");
```

Comment:
The parseDomain("") method only parses the domain given in the file path if the domain has not already been parsed. Otherwise it returns a pointer to the already parsed VAL domain object, and the file path is not necessary.

A better thing would be to have "hasAlreadyParsed" and "getLastDomainParsed" methods.(Maybe we can open a new issue for this, and add in future).

Reply:
But the domain in the PDDLParsers is already public (we may need to discuss that at some point, but all 3 parsers should do the same), so why not just calling domain_parser.domain as done in other places such as line 38/64/90, better if it has an already parsed check like the one in line 20?

Reply:
Agreed. Will push a change.